src/main/java/appeng/items/storage/ItemReadOnlyStorageCell.java

```diff
@@ -0,0 +1,240 @@
+ *
+ * Applied Energistics 2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.items.storage;
+
+
+import appeng.api.AEApi;
+import appeng.api.config.FuzzyMode;
+import appeng.api.config.IncludeExclude;
+import appeng.api.exceptions.MissingDefinition;
+import appeng.api.implementations.items.IItemGroup;
+import appeng.api.implementations.items.IReadOnly;
+import appeng.api.implementations.items.IStorageCell;
+import appeng.api.implementations.items.IUpgradeModule;
+import appeng.api.storage.ICellInventory;
+import appeng.api.storage.ICellInventoryHandler;
+import appeng.api.storage.IMEInventoryHandler;
+import appeng.api.storage.StorageChannel;
+import appeng.api.storage.data.IAEItemStack;
+import appeng.api.storage.data.IItemList;
+import appeng.core.AEConfig;
+import appeng.core.features.AEFeature;
+import appeng.core.localization.GuiText;
+import appeng.items.AEBaseItem;
+import appeng.items.contents.CellConfig;
+import appeng.items.contents.CellUpgrades;
+import appeng.items.materials.MaterialType;
+import appeng.util.InventoryAdaptor;
+import appeng.util.Platform;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.InventoryPlayer;
+import net.minecraft.inventory.IInventory;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.util.ActionResult;
+import net.minecraft.util.EnumActionResult;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.EnumHand;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import java.util.List;
+import java.util.Set;
+
+
+public final class ItemReadOnlyStorageCell extends AEBaseItem implements IStorageCell, IItemGroup, IReadOnly
+{
+	private final int totalBytes;
+	private final int perType;
+	private final double idleDrain;
+
+	public ItemReadOnlyStorageCell()
+	{
+		this.setMaxStackSize( 1 );
+		this.totalBytes = 64 * 1024;
+		this.idleDrain = 2.0;
+		this.perType = 512;
+	}
+
+	@Override
+	public void addCheckedInformation( final ItemStack stack, final EntityPlayer player, final List<String> lines, final boolean displayMoreInfo )
+	{
+		final IMEInventoryHandler<?> inventory = AEApi.instance().registries().cell().getCellInventory( stack, null, StorageChannel.ITEMS );
+
+		if( inventory instanceof ICellInventoryHandler )
+		{
+			final ICellInventoryHandler handler = (ICellInventoryHandler) inventory;
+			final ICellInventory cellInventory = handler.getCellInv();
+
+			if( cellInventory != null )
+			{
+				if( getReadOnly( stack ) )
+				{
+					if( cellInventory.getUsedBytes() == 0 )
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)